data/build.js

```diff
@@ -74,6 +74,73 @@ function createTreeSorter(name) {
   };
 }
 
+async function buildScientificNameToImageMap(names) {
+  function partition(arr, size) {
+    const numGroups = Math.ceil(arr.length / size);
+    const groups = new Array(numGroups).fill(null);
+
+    return groups.map((_, i) => arr.slice(i * size, i * size + size));
+  }
+
+  function buildRequestURL(titles) {
+    const BASE_URL = 'https://en.wikipedia.org/w/api.php';
+    const requestURL = new URL(BASE_URL);
+    requestURL.search = new URLSearchParams({
+      action: 'query',
+      format: 'json',
+      prop: 'pageimages',
+      piprop: 'thumbnail',
+      pithumbsize: '600',
+      titles,
+    });
+
+    return requestURL;
+  }
+
+  function getRequestURLs(data) {
+    const MAX_TITLES_PER_QUERY = 50;
+    const partitions = partition(data, MAX_TITLES_PER_QUERY);
+    const queryTitles = partitions.map((titles) => titles.join('|'));
+    const queries = queryTitles.map((titles) => buildRequestURL(titles));
+    return queries.map((query) => query.href);
+  }
+
+  async function getResponseJSON(url) {
+    const response = await fetch(url);
+    return response.json();
+  }
+
+  async function getImages(requestURLs) {
+    const images = {};
+    const responsesMap = requestURLs.map((url) => getResponseJSON(url));
+    const responses = await Promise.all(responsesMap);
+    const allPages = responses.reduce((accumulator, response) => {
+      const { query: { pages = [] } = {} } = response;
+      return accumulator.concat(...Object.values(pages));
+    }, []);
+
+    Object.values(allPages).forEach((page) => {
+      const { title, thumbnail: { source } = {} } = page;
+      images[title] = source;
+    });
+
+    const allNormalized = responses.reduce((accumulator, response) => {
+      const { query: { normalized = [] } = {} } = response;
+      return accumulator.concat(...normalized);
+    }, []);
```

Comment:
Wiki API automatically normalizes some tree names. We want the original names from our json files.

Reply:
Pretty much fixes typos. The Wiki API formats the tree names we give it. Since I'm keying into the imageURLs object using the names from our json data files, I have to convert the normalized names back to the original from our data. Otherwise, I would be trying to key into something that can never exist.


Example of a response from the API:
```
{
    "batchcomplete": "",
    "query": {
        "normalized": [
            {
                "from": "coast live oak (Quercus agrifolia)",
                "to": "Coast live oak (Quercus agrifolia)"
            },
            {
                "from": "Pyracantha  'Santa Cruz'",
                "to": "Pyracantha 'Santa Cruz'"
            },
            {
                "from": "ulmus x frontier Frontier Hybrid Elm",
                "to": "Ulmus x frontier Frontier Hybrid Elm"
            },
            {
                "from": "arbutus marina",
                "to": "Arbutus marina"
            }
        ],
        "pages": {
            "-1": {
                "ns": 0,
                "title": "Coast live oak (Quercus agrifolia)",
                "missing": ""
            },
            "-2": {
                "ns": 0,
                "title": "Pyracantha 'Santa Cruz'",
                "missing": ""
            },
            "-3": {
                "ns": 0,
                "title": "Ulmus x frontier Frontier Hybrid Elm",
                "missing": ""
            },
            "-4": {
                "ns": 0,
                "title": "Arbutus marina",
                "missing": ""
            }
        }
    }
}
```